### ISSUE TITLE
Implement outer to inner join conversion rule

### ISSUE THREAD

**Issue by aalexandrov** [MaterializeInc/materialize#10239]

**Comment by wangandi** [comment]
The way I see it, the description of this issue would be best implemented as at least four separate rewrites: 
* `InlineCSE`: applied PreOrder. The rule is: make clones of a child box if the child box is not a Get and has two or more parents.
* `PredicatePushdown`: applied PreOrder.
* `OuterJoinSimplication`: applied PreOrder. Is just implemented as a variant of Algorithm A. Does not clone.
* `FuseIdenticalParents`: applied PostOrder. The rule is: if a box has two or more identical parent boxes, move all quantifiers refer to exactly one of those boxes. 

How I envision these four rewrites working together:

`InlineCSE` will turn the graph below on the left to the graph below on the right no matter what the boxes in the graph are.
```
   B0                 B0
  /  \               /  \
 /    \             /    \
B1    B2    ==>    B1    B2
|     |            |     |
B3----/            B3    B3'
|                  |     |
B4                 B4    B4'
|                  |     |
B5                 B5    B5'
|                  |     |
B6                 B6----´
```

Suppose in the diagram, `B4` is an outer join, `B1` rejects nulls produced by `B4`, but `B2` does not. Now that `B3`-`B5` have been cloned, we can push the information that `B1` rejects nulls down to `B4`, perhaps resulting in a new box `B7` existing between `B3` and `B4`.

```
   B0                 B0
  /  \               /  \
 /    \             /    \
B1    B2    ==>    B1    B2
|     |            |     |
B3    B3'          B3    B3'
|     |            |     |
B4    B4'          B7    |
|     |            |     |
B5    B5'          B4    B4'
|     |            |     |
B6----´            B5    B5'
                   |     |
                   B6----´
```

`OuterJoinSimplication` turns `B4` into `B4"`, where at least one quantifier has changed from preserving to non-preserving.  

```
   B0                 B0
  /  \               /  \
 /    \             /    \
B1    B2    ==>    B1    B2
|     |            |     |
B3    B3'          B3    B3'
|     |            |     |
B7    |            B7    |
|     |            |     |
B4    B4'          B4"   B4'
|     |            |     |
B5    B5'          B5    B5'
|     |            |     |
B6----´            B6----´
```

Since `B5` and `B5'` have remained the same, `FuseIdenticalParents` turns them back into one box.

```
   B0                 B0
  /  \               /  \
 /    \             /    \
B1    B2    ==>    B1    B2
|     |            |     |
B3    B3'          B3    B3'
|     |            |     |
B7    |            B7    |
|     |            |     |
B4    B4'          B4"   B4'
|     |            |     |
B5    B5'          B5----´  
|     |            | 
B6----´            B6
```

**Comment by aalexandrov** [comment]
The approach sounds reasonable to me, with the following remarks:

1. I prefer to have more straight-forward names for the first and final transform that also indicate that they are inverse. For example:
    1. `InlineCSE` ⇒ `InlineCommonSubexpressions`.
    1. `FuseIdenticalParents` ⇒ `EliminateCommonSubexpressions`.
1. I am not sure why you propose to not clone `Get` nodes as part of `InlineCSE`. This is fine for the current rewrite, but it might be for other rewrites or attributes that expect a model that is a tree.
1. With the current framework rather than having a separate rule (`PredicatePushdown`) we should have a derived attribute that either depends on or extends `RejectedNulls`. The attribute should be stating as an invariant that it expects a QGM model that is a tree and should accumulate the sets of rejected nulls for the (unique) ancestor chain of the current box `b` in a single set, handling remapping in the process (so you correctly state that it needs to state `PreOrder` visit strategy).

Overall, I am mildly depressed that after all the talk and examples of the DAG-like structure of QGM simplifying our lives the very first rule that we try to implement bites us like that... Oh, well!

**Comment by aalexandrov** [comment]
Maybe one way to scope the effort here is to have an initial version with a more restricted `check` that assumes that the ancestor path is unique. This won't be as complete as the other approach but is still sound.

**Comment by wangandi** [comment]
> 2. I am not sure why you propose to not clone Get nodes as part of InlineCSE. This is fine for the current rewrite, but it might be for other rewrites or attributes that expect a model that is a tree.

I think of `Get`s as 1) immutable 2) leaf nodes. So it should not ever be necessary to clone `Get` nodes. Note that in our current HIR -> QGM code, all branches of code getting the same view/source point to the same `Get` box from the beginning.  https://github.com/MaterializeInc/materialize/blob/31da762d4fdfe37cfd40ad0ad7fe839e4ce6304e/src/sql/src/query_model/hir.rs#L68

I do not think that any rewrite or attribute should think of the query graph model as a tree given the fact that we support CTEs and lateral joins (and possibly recursive CTEs in the future). 

To me, the main benefit of not cloning `Get` would be that it would allow `EliminateCommonSubexpressions` to work with  the rewrite engine. I think that all rewrites should be able to inspect a single box and determine if the area of the graph near the box needs to be changed. If we were to clone `Get`, the first thing `EliminateCommonSubexpressions` would have to do is do a global scan of the graph to find all the `Get` leaves and fuse the `Get`s that point to the same source/view. 

Obviously, the version of `EliminateCommonSubexpressions` that does not do a global scan is less powerful than `src::transform::RelationCSE`, but I think that's fine because:
1) We should have some kind of `ConstantFolding` transform eventually that would collapse a subgraph whose leaves are just `Constant`s and `TableFunction` down into one box. 
2) We are not deleting `src::transform::RelationCSE` any time soon.

> 3. I am not sure why you think that we might need to introduce extra boxes (such as B7). Can you elaborate?

The short answer is that in the `PredicatePushdown` section of the QGM handover notes  it reads,
> Note that a predicate from a Select box can be pushed to another Select Box. If there isn’t a Select Box at the end of the path the predicate can be pushed down through, we must add a new one.

The long answer is an example:
Suppose `B1` is a Select box that contained the predicate `t1.f1 > t2.f1`.
Suppose `B3` is a Grouping box grouping by `(t1.f1, t2.f1)`
Suppose `B4` is an Outer Join t1 left join t2 on `<join conditions>`. 

If you push down the predicate `t1.f1 > t2.f1`, it can't go inside the `B3` box because Grouping boxes do not have a `predicates` field. It can't go below `B4` for obvious reasons. It can't go into the `predicates` field of `B4` because that would change the predicates in the on clause of the join. Thus, it can only go inside a new select box between `B3` and `B4`.

**Comment by wangandi** [comment]
> 4. With the current framework rather than having a separate rule (PredicatePushdown) we should have a derived attribute that either depends on or extends RejectedNulls. The attribute should be stating as an invariant that it expects a QGM model that is a tree and should accumulate the sets of rejected nulls for the (unique) ancestor chain of the current box b in a single set, handling remapping in the process (so you correctly state that it needs to state PreOrder visit strategy).

Yeah, we can just worry about `PredicatePushdown` later. The derived attribute only needs to assume that the box has a unique ancestor chain, no? As opposed to the whole QGM model being a tree?

> One way to narrow down the scope for the initial version is to have a more restricted check that assumes that the ancestor path is unique. This won't be as complete as the other approach but is still sound. As we discussed offline, for the first iteration the accumulated version of RejectedNulls described in my previous comment can also be computed ad-hoc as part of the OuterJoinSimplication rule.


Reasonable.

**Comment by aalexandrov** [comment]
OK, so point (3) is referring to something that we need to do if we want to do `PredicatePushdown`. I don't think we need that part in order to simplify outer joins. In general, if `B1` has a predicate `t1.f1 > t2.f1`, and the child box `B2` is grouping by `t1.f1` only, we should still be able to infer that `B3` can be simplified even if we cannot push down the predicate to an immediate parent `Select` box of `B3`.

**Comment by aalexandrov** [comment]
> I think that all rewrites should be able to inspect a single box and determine if the area of the graph near the box needs to be changed

At the first glance this seems to enforce your previous assumption

> I do not think that any rewrite or attribute should think of the query graph model as a tree given the fact that we support CTEs and lateral joins (and possibly recursive CTEs in the future).

but I think there is a way to relax the latter while still offering an API that is expressive enough for the former. Again, this most probably will come with the price of giving up explicit mutability.

**Comment by aalexandrov** [comment]
Note to self: with the current draft PR (#10509) this rewrite will take a quadratic number of tree traversals in the depth of the tree in the worst case (at every node, we traverse the entire subtree). We should open a follow-up issue to fix that by modeling the stack variables in the `check` method as an additional top-down derived variable.
